### DISCUSSION TITLE
Derivative Properties Always Return Zero

### DISCUSSION THREAD

**Discussion by Changster101** [reaktoro/reaktoro#386] [Q&A]
### Description

I am currently testing how to find derivative properties (i.e. specificVolumeT and specificVolumeP) after solving a chemical state however the functions always return 0. I've tested incrementing the temperature and condition constraints manually, however the derivative is still not calculated.  Could you explain how these functions work, and how I can use them to get, for example, the temperature derivative of specific volume, since I can see that between the initial state and final state the specific volume does change.

Here's an example code I've been testing with:

### Example

```


if __name__ == "__main__":
      db = PhreeqcDatabase("frezchem.dat")
      mineralPhase = MineralPhases()
      solution = AqueousPhase("H2O")

    system = ChemicalSystem(db, solution, mineralPhase)

    state = ChemicalState(system)
    state.add("H2O", 0.1, "g")
    state.temperature(300, "K")
    state.pressure(100, "bar")
    props = ChemicalProps(state)

    print("INITIAL STATE")
    print(state)
    print(props.density())
    print(props.specificVolume())

    specs = EquilibriumSpecs(system)
    specs.pressure()
    specs.temperature()

    solver = EquilibriumSolver(specs)

    conditions = EquilibriumConditions(specs)
    conditions.pressure(300, "bar")
    conditions.temperature(200, "K")

    solver.solve(state, conditions)
    props.update(state)

    print("FINAL STATE")
    print(props.specificVolume())
    print(props.specificVolumeT())
    print(props.specificVolumeP())
    print(props.molarVolumeT())
    print(props.volumeT())
```

### What is your operating system?

macOS

**Comment by allanleal** [comment]
The method `ChemicalProps.volumeT` computes the temperature derivative of the total volume of the state. At 200K and 300bar, the initial amount of H2O is stable as ice. In PHREEQC databases, molar volumes of solids (e.g., ice) are constant. Therefore, a zero temperature-derivative for volume, in these conditions, is expected.

However, if we increase the temperature to 300K, at which H2O is stable as liquid, `ChemicalProps.volumeT` is still producing zero because temperature-derivative for molar volumes of these phases are currently available for gases only. When `volumeT` asks for the molar volumes of the phases internally, it gets zero for both the solid and the aqueous phases.

**Reply by Changster101** [reply]
Hi @allanleal, thank you for the response! I see, so the derivative properties are only available for the gaseous phase. One of my goals for a program I am writing is to calculate the thermal expansion coefficient of water for a given temperature and pressure. One derivation of this equation is to multiply the reciprocal of specific volume with its derivative with respect to temperature (linked the equation below). Since I can only get the specific volume derivative with respect to temperature, would you suggest 'manually' computing the derivative (i.e. incrementing T slightly and calculating the corresponding change in specific volume)? Or is there another way I can obtain the thermal expansion coefficient?


<img width="142" alt="Screenshot 2024-06-17 at 8 57 28 AM" src="https://github.com/reaktoro/reaktoro/assets/68867402/8d0a62a3-d102-4a32-8021-58e8445af50a">

**Reply by allanleal** [reply]
Hi @Changster101 - thanks for providing more details about what you intend to do. Would it suffice to use the equation of state of water (e.g., from Wager & Pruss, 2002) to compute these derivatives? The code below illustrates all thermodynamic properties available once the EOS is evaluated:

~~~py
import reaktoro as rkt

T = 300.0         # temperature in K
P = 100.0 * 1e+5  # pressure in Pa
wprops = rkt.waterThermoPropsWagnerPruss(T, P, StateOfMatter.Liquid)

print("The temperature of water (in K)")
print(wprops.T)
 
print("The specific volume of water (in m3/kg)")
print(wprops.V)
 
print("The specific entropy of water (in J/(kg*K))")
print(wprops.S)
 
print("The specific Helmholtz free energy of water (in J/kg)")
print(wprops.A)
 
print("The specific internal energy of water (in J/kg)")
print(wprops.U)
 
print("The specific enthalpy of water (in J/kg)")
print(wprops.H)
 
print("The specific Gibbs free energy of water (in J/kg)")
print(wprops.G)
 
print("The specific isochoric heat capacity of water (in J/(kg*K))")
print(wprops.Cv)
 
print("The specific isobaric heat capacity of water (in J/(kg*K))")
print(wprops.Cp)
 
print("The specific density of water (in kg/m3)")
print(wprops.D)
 
print("The first-order partial derivative of density with respect to temperature (in (kg/m3)/K)")
print(wprops.DT)
 
print("The first-order partial derivative of density with respect to pressure (in (kg/m3)/Pa)")
print(wprops.DP)
 
print("The second-order partial derivative of density with respect to temperature (in (kg/m3)/(K*K))")
print(wprops.DTT)
 
print("The second-order partial derivative of density with respect to temperature and pressure (in (kg/m3)/(K*Pa))")
print(wprops.DTP)
 
print("The second-order partial derivative of density with respect to pressure (in (kg/m3)/(Pa*Pa))")
print(wprops.DPP)
 
print("The pressure of water (in Pa)")
print(wprops.P)
 
print("The first-order partial derivative of pressure with respect to temperature (in Pa/K)")
print(wprops.PT)
 
print("The first-order partial derivative of pressure with respect to density (in Pa/(kg/m3))")
print(wprops.PD)
 
print("The second-order partial derivative of pressure with respect to temperature (in Pa/(K*K))")
print(wprops.PTT)
 
print("The second-order partial derivative of pressure with respect to temperature and density (in Pa/(K*kg/m3))")
print(wprops.PTD)
 
print("The second-order partial derivative of pressure with respect to density (in Pa/((kg/m3)*(kg/m3)))")
print(wprops.PDD)

# Output

# The temperature of water (in K)
# 300
# The specific volume of water (in m3/kg)
# 0.000999046
# The specific entropy of water (in J/(kg*K))
# 390.29
# The specific Helmholtz free energy of water (in J/kg)
# -5343.23
# The specific internal energy of water (in J/kg)
# 111744
# The specific enthalpy of water (in J/kg)
# 121734
# The specific Gibbs free energy of water (in J/kg)
# 4647.22
# The specific isochoric heat capacity of water (in J/(kg*K))
# 4098.4
# The specific isobaric heat capacity of water (in J/(kg*K))
# 4153.63
# The specific density of water (in kg/m3)
# 1000.96
# The first-order partial derivative of density with respect to temperature (in (kg/m3)/K)
# -0.284791
# The first-order partial derivative of density with respect to pressure (in (kg/m3)/Pa)
# 4.39721e-07
# The second-order partial derivative of density with respect to temperature (in (kg/m3)/(K*K))
# -0.00863079
# The second-order partial derivative of density with respect to temperature and pressure (in (kg/m3)/(K*Pa))
# -1.09521e-09
# The second-order partial derivative of density with respect to pressure (in (kg/m3)/(Pa*Pa))
# -9.15647e-16
# The pressure of water (in Pa)
# 1e+07
# The first-order partial derivative of pressure with respect to temperature (in Pa/K)
# 647664
# The first-order partial derivative of pressure with respect to density (in Pa/(kg/m3))
# 2.27417e+06
# The second-order partial derivative of pressure with respect to temperature (in Pa/(K*K))
# 23727.6
# The second-order partial derivative of pressure with respect to temperature and density (in Pa/(K*kg/m3))
# 8731.36
# The second-order partial derivative of pressure with respect to density (in Pa/((kg/m3)*(kg/m3)))
# 10769.6
~~~

The method `waterThermoPropsWagnerPruss` produces a [`WaterThermoProps`](https://reaktoro.org/api/structReaktoro_1_1WaterThermoProps.html) object.

I think you could use `wprops.DT` in the process to compute the temperature-derivative of volume, where `wprops.DT` is $\partial \rho/\partial T$ at constant $P$.
